In [3]:
# importing the data set
import pandas as pd 

In [5]:
df=pd.read_csv("employee_attrition_prediction_dataset.csv")

In [7]:
df.head()

,Employee_ID,Age,Department,Job_Level,Monthly_Salary,Years_At_Company,Job_Satisfaction,Work_Life_Balance,Overtime,Training_Hours,Performance_Rating,Remote_Work_Available,Left_Company
0,1,32.0,HR,Entry,3762.0,7,3.0,5,Yes,18,4,Yes,No
1,2,47.0,Marketing,Entry,14081.0,8,2.0,2,No,39,1,No,No
2,3,22.0,IT,Intermediate,8804.0,4,5.0,5,No,47,4,Yes,No
3,4,25.0,HR,Intermediate,13462.0,8,3.0,1,Yes,26,1,No,No
4,5,35.0,IT,Entry,13677.0,9,1.0,1,No,6,5,Yes,No


In [9]:
with open("employee_attrition_prediction_dataset.csv", "r", encoding="utf-8") as f:
    print(f.readline())

Employee_ID,Age,Department,Job_Level,Monthly_Salary,Years_At_Company,Job_Satisfaction,Work_Life_Balance,Overtime,Training_Hours,Performance_Rating,Remote_Work_Available,Left_Company



In [11]:
#changing the format 
with open("employee_attrition_prediction_dataset.csv", "r", encoding="utf-8-sig") as f:
    lines = [line.strip().strip('"') for line in f]

with open("employee_attrition_prediction_dataset.csv", "w", encoding="utf-8") as f:
    for line in lines:
        f.write(line + "\n")

In [13]:
import pandas as pd
df=pd.read_csv("employee_attrition_prediction_dataset.csv")
df.head()

,Employee_ID,Age,Department,Job_Level,Monthly_Salary,Years_At_Company,Job_Satisfaction,Work_Life_Balance,Overtime,Training_Hours,Performance_Rating,Remote_Work_Available,Left_Company
0,1,32.0,HR,Entry,3762.0,7,3.0,5,Yes,18,4,Yes,No
1,2,47.0,Marketing,Entry,14081.0,8,2.0,2,No,39,1,No,No
2,3,22.0,IT,Intermediate,8804.0,4,5.0,5,No,47,4,Yes,No
3,4,25.0,HR,Intermediate,13462.0,8,3.0,1,Yes,26,1,No,No
4,5,35.0,IT,Entry,13677.0,9,1.0,1,No,6,5,Yes,No


In [15]:
df.shape

(366, 13)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Employee_ID            366 non-null    int64  
 1   Age                    360 non-null    float64
 2   Department             359 non-null    object 
 3   Job_Level              359 non-null    object 
 4   Monthly_Salary         360 non-null    float64
 5   Years_At_Company       366 non-null    int64  
 6   Job_Satisfaction       360 non-null    float64
 7   Work_Life_Balance      366 non-null    int64  
 8   Overtime               366 non-null    object 
 9   Training_Hours         366 non-null    int64  
 10  Performance_Rating     366 non-null    int64  
 11  Remote_Work_Available  360 non-null    object 
 12  Left_Company           366 non-null    object 
dtypes: float64(3), int64(5), object(5)
memory usage: 37.3+ KB


In [19]:
#Business Problem:
business_problem= "Can we predict whether an employee will leave the company based on their work-related and demographic information?"

In [21]:
#Machine learning problem:
ml_problem = "This is a classification problem because we want to predict whether the emplyoee will leave the company or not"

In [23]:
# Prediction goal:

prediction_goal = "Predict whether an employee will leave the company: Yes or No."

# Display the problem definition
print("Business Problem:")
print(business_problem)

print("\nMachine Learning Problem:")
print(ml_problem)
print("\nPrediction Goal:")
print(prediction_goal)

Business Problem:
Can we predict whether an employee will leave the company based on their work-related and demographic information?

Machine Learning Problem:
This is a classification problem because we want to predict whether the emplyoee will leave the company or not

Prediction Goal:
Predict whether an employee will leave the company: Yes or No.


In [25]:
df.duplicated().sum()

6

In [27]:
#Checking if there are duplicated rows
df[df.duplicated(keep=False)]

,Employee_ID,Age,Department,Job_Level,Monthly_Salary,Years_At_Company,Job_Satisfaction,Work_Life_Balance,Overtime,Training_Hours,Performance_Rating,Remote_Work_Available,Left_Company
155,156,26.0,HR,Senior,5148.0,10,3.0,5,No,54,3,No,No
176,177,54.0,NaN,Entry,14566.0,10,5.0,2,No,48,1,No,No
181,182,46.0,Sales,Intermediate,5363.0,5,5.0,1,No,68,4,Yes,No
219,220,44.0,HR,Intermediate,6076.0,19,1.0,2,No,70,4,No,No
312,313,39.0,IT,NaN,14571.0,8,4.0,2,No,10,5,No,No
320,321,61.0,Operations,Intermediate,14526.0,12,4.0,3,Yes,36,4,Yes,No
360,313,39.0,IT,NaN,14571.0,8,4.0,2,No,10,5,No,No
361,220,44.0,HR,Intermediate,6076.0,19,1.0,2,No,70,4,No,No
362,182,46.0,Sales,Intermediate,5363.0,5,5.0,1,No,68,4,Yes,No
363,156,26.0,HR,Senior,5148.0,10,3.0,5,No,54,3,No,No


In [29]:
#removing duplicated rows
df = df.drop_duplicates(subset=df.columns)

In [31]:
df.duplicated().sum()

0

In [33]:
df.drop

<bound method DataFrame.drop of      Employee_ID   Age Department     Job_Level  Monthly_Salary  \
0              1  32.0         HR         Entry          3762.0   
1              2  47.0  Marketing         Entry         14081.0   
2              3  22.0         IT  Intermediate          8804.0   
3              4  25.0         HR  Intermediate         13462.0   
4              5  35.0         IT         Entry         13677.0   
..           ...   ...        ...           ...             ...   
355          356  27.0         HR         Entry          6482.0   
356          357  27.0         HR  Intermediate          9220.0   
357          358  28.0  Marketing       Manager          3071.0   
358          359  39.0      Sales        Senior         11186.0   
359          360  38.0         HR        Senior          8245.0   

     Years_At_Company  Job_Satisfaction  Work_Life_Balance Overtime  \
0                   7               3.0                  5      Yes   
1                   8

In [39]:
# removing emplyoee id because we won't need it for the prediction
df=df.drop(columns=['Employee_ID'])

In [57]:
df.isnull().sum()

Age                      6
Department               6
Job_Level                6
Monthly_Salary           6
Years_At_Company         0
Job_Satisfaction         6
Work_Life_Balance        0
Overtime                 0
Training_Hours           0
Performance_Rating       0
Remote_Work_Available    6
Left_Company             0
dtype: int64

In [123]:
#Handling numerical missing values 
df["Age"]= df["Age"].fillna(df["Age"].median())
df["Monthly_Salary"]= df["Monthly_Salary"].fillna(df["Monthly_Salary"].median())
df["Job_Satisfaction"]= df["Job_Satisfaction"].fillna(df["Job_Satisfaction"].median())

In [125]:
#Handling categorical values 
df["Department"] = df["Department"].fillna(df["Department"].mode()[0])
df["Job_Level"]  = df["Job_Level"].fillna(df["Job_Level"].mode()[0])
df["Remote_Work_Available"]  = df["Remote_Work_Available"].fillna(df["Remote_Work_Available"].mode()[0])



In [127]:
df.isnull().sum()

Age                      0
Department               0
Job_Level                0
Monthly_Salary           0
Years_At_Company         0
Job_Satisfaction         0
Work_Life_Balance        0
Overtime                 0
Training_Hours           0
Performance_Rating       0
Remote_Work_Available    0
Left_Company             0
dtype: int64

In [129]:
#Features and target variable
X=df.drop(columns=['Left_Company'])
y=df['Left_Company']

In [131]:
X.head()

,Age,Department,Job_Level,Monthly_Salary,Years_At_Company,Job_Satisfaction,Work_Life_Balance,Overtime,Training_Hours,Performance_Rating,Remote_Work_Available
0,32.0,HR,Entry,3762.0,7,3.0,5,Yes,18,4,Yes
1,47.0,Marketing,Entry,14081.0,8,2.0,2,No,39,1,No
2,22.0,IT,Intermediate,8804.0,4,5.0,5,No,47,4,Yes
3,25.0,HR,Intermediate,13462.0,8,3.0,1,Yes,26,1,No
4,35.0,IT,Entry,13677.0,9,1.0,1,No,6,5,Yes


In [133]:
y.head()

0    No
1    No
2    No
3    No
4    No
Name: Left_Company, dtype: object

In [135]:
#Preprocessing 
from sklearn.compose import ColumnTransformer 
from sklearn.preprocessing import OneHotEncoder, StandardScaler

In [137]:
# Converting the target variable from text labels to numbers
y = y.map({"No":0, "Yes":1})

In [139]:
#numerical features to be scaled 
numerical_features =[
    "Age",
    "Monthly_Salary",
    "Years_At_Company",
    "Job_Satisfaction",
    "Work_Life_Balance",
    "Training_Hours",
    "Performance_Rating"
]
    
    
    
    

In [141]:
# Categorical features 
categorical_features = [
    "Department",
    "Job_Level",
    "Overtime"
]
    
                

In [143]:
# StandardScaler will be applied to numerical columns
# OneHotEncoder will be applied to categorical columns
preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numerical_features),
        ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_features)
    ]
)

In [145]:
# Display the first 5 rows of the feature matrix
print("Features before preprocessing:")

display(X.head())

Features before preprocessing:


,Age,Department,Job_Level,Monthly_Salary,Years_At_Company,Job_Satisfaction,Work_Life_Balance,Overtime,Training_Hours,Performance_Rating,Remote_Work_Available
0,32.0,HR,Entry,3762.0,7,3.0,5,Yes,18,4,Yes
1,47.0,Marketing,Entry,14081.0,8,2.0,2,No,39,1,No
2,22.0,IT,Intermediate,8804.0,4,5.0,5,No,47,4,Yes
3,25.0,HR,Intermediate,13462.0,8,3.0,1,Yes,26,1,No
4,35.0,IT,Entry,13677.0,9,1.0,1,No,6,5,Yes


In [155]:
# Display the first 5 values of the target variable
print("Target after converting Yes/No to 1/0:")

display(y.head())

Target after converting Yes/No to 1/0:


0    0
1    0
2    0
3    0
4    0
Name: Left_Company, dtype: int64

In [157]:
from sklearn.model_selection import train_test_split

In [159]:
#splitting data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42,
                                                    stratify=y)

In [161]:
# Display the shape of the training and testing sets

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (288, 11)
X_test shape: (72, 11)
y_train shape: (288,)
y_test shape: (72,)


In [163]:
# Display the first 5 rows of the training features

print("Training features:")
display(X_train.head())

# Display the first 5 rows of the testing features

print("Testing features:")
display(X_test.head())

Training features:


,Age,Department,Job_Level,Monthly_Salary,Years_At_Company,Job_Satisfaction,Work_Life_Balance,Overtime,Training_Hours,Performance_Rating,Remote_Work_Available
251,36.0,Finance,Entry,3893.0,9,1.0,4,No,72,5,No
187,39.0,Marketing,Intermediate,8341.0,9,4.0,5,Yes,8,5,Yes
75,53.0,Operations,Entry,4380.0,12,1.0,5,No,36,5,Yes
152,35.0,Operations,Manager,5181.0,19,2.0,4,Yes,3,1,Yes
255,64.0,Marketing,Entry,8813.5,2,5.0,2,Yes,5,3,Yes


Testing features:


,Age,Department,Job_Level,Monthly_Salary,Years_At_Company,Job_Satisfaction,Work_Life_Balance,Overtime,Training_Hours,Performance_Rating,Remote_Work_Available
329,38.0,IT,Entry,11762.0,5,2.0,5,Yes,48,1,Yes
135,48.0,Marketing,Entry,3854.0,15,2.0,1,No,40,1,No
209,34.0,Finance,Intermediate,14723.0,13,4.0,5,No,51,5,Yes
268,42.0,Marketing,Intermediate,7159.0,6,1.0,4,No,67,2,Yes
118,60.0,Finance,Entry,5716.0,19,4.0,2,Yes,7,3,Yes


In [165]:
# Fit the preprocessor only on the training data
# This prevents data leakage from the test set

X_train_processed = preprocessor.fit_transform(X_train)

# Transform the test data using the same preprocessor
# We do not fit the preprocessor again on the test data

X_test_processed = preprocessor.transform(X_test)

In [167]:
# Display the shape before and after preprocessing

print("Before preprocessing:")
print("X_train:", X_train.shape)
print("X_test:", X_test.shape)

print("\nAfter preprocessing:")
print("X_train_processed:", X_train_processed.shape)
print("X_test_processed:", X_test_processed.shape)

Before preprocessing:
X_train: (288, 11)
X_test: (72, 11)

After preprocessing:
X_train_processed: (288, 19)
X_test_processed: (72, 19)


In [169]:
#Basline Model 
# Import Logistic Regression from scikit-learn
# Logistic Regression is a simple and commonly used classification model
from sklearn.linear_model import LogisticRegression

In [171]:
# Create a Logistic Regression model
# max_iter is increased to make sure the model has enough iterations to train properly

baseline_model = LogisticRegression(max_iter=1000)

In [173]:
# Train the baseline model using the processed training data
# X_train_processed contains the preprocessed features
# y_train contains the correct target labels

baseline_model.fit(X_train_processed, y_train)

LogisticRegression(max_iter=1000)

In [175]:
# Use the trained model to make predictions on the processed test data
# The model has not seen X_test_processed during training

y_pred = baseline_model.predict(X_test_processed)

In [179]:
# 0 means the employee left the company
# 1 means the employee stayed in the company
print(y_pred)

[0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0]


In [181]:
#  Evaluating the Baseline Model

# Import evaluation metrics from scikit-learn
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score, recall_score, f1_score
from sklearn.metrics import classification_report

In [183]:
# Calculate the accuracy of the baseline model
# Accuracy shows the percentage of correct predictions

accuracy = accuracy_score(y_test, y_pred)

# Display the accuracy score
print("Accuracy:", accuracy)

Accuracy: 0.9027777777777778


In [185]:
# Create the confusion matrix
# It compares the actual values with the predicted values

cm = confusion_matrix(y_test, y_pred)

# Display the confusion matrix
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[64  2]
 [ 5  1]]


In [187]:
# Display the confusion matrix as a table

import pandas as pd

cm_table = pd.DataFrame(
    cm,
    index=["Actual No", "Actual Yes"],
    columns=["Predicted No", "Predicted Yes"]
)

cm_table

,Predicted No,Predicted Yes
Actual No,64,2
Actual Yes,5,1


In [189]:
# Calculate precision
# Precision measures how many predicted Yes values were actually Yes
precision = precision_score(y_test, y_pred, zero_division=0)

# Calculate recall
# Recall measures how many actual Yes values were correctly predicted
recall = recall_score(y_test, y_pred, zero_division=0)

# Calculate F1-score
# F1-score balances precision and recall
f1 = f1_score(y_test, y_pred, zero_division=0)

# Display the evaluation results
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Precision: 0.3333333333333333
Recall: 0.16666666666666666
F1-score: 0.2222222222222222


In [191]:
# Simple model evaluation summary

print("Model Evaluation Summary")
print("------------------------")
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

print("\nInterpretation:")
print("The baseline model gives us an initial performance result.")
print("We can use this result as a benchmark for comparing more advanced models later.")

Model Evaluation Summary
------------------------
Accuracy: 0.9027777777777778
Precision: 0.3333333333333333
Recall: 0.16666666666666666
F1-score: 0.2222222222222222

Interpretation:
The baseline model gives us an initial performance result.
We can use this result as a benchmark for comparing more advanced models later.


In [193]:
#Interpreting Results

## Task 2


# Model Explanation

## What my model is trying to predict
My model is trying to predict whether an employee will leave the company or not.

## What dataset I used
I used the employee dataset that was provided by the professor.

## What features and target I selected

### Features
- Age
- Department
- Job_Level
- Monthly_Salary
- Years_At_Company
- Job_Satisfaction
- Work_Life_Balance
- Overtime
- Training_Hours
- Performance_Rating
- Remote_Work_Available

### Target
- Left_Company

## What result I obtained

### Model Evaluation Summary
- Accuracy: 0.9027777777777778
- Precision: 0.3333333333333333
- Recall: 0.16666666666666666
- F1-score: 0.2222222222222222
  
- **Accuracy:** The model correctly predicted employee outcomes about 90% of the time overall.

- **Precision:** When the model predicted that an employee would leave, it was correct about 33% of the time.

- **Recall:** The model only identified about 17% of the employees who actually left the company.

- **F1-score:** The low F1-score indicates that the model had difficulty balancing precision and recall effectively.

## One limitation of my model or dataset
One limitation of my model is that the dataset may be too small and imbalanced, which can affect the model’s ability to correctly predict employees who leave the company. Another limitation is the handling of missing values, since improper handling of missing data can reduce model performance and prediction reliability.